src/clojupyter/kernel/comm_atom.clj

```diff
@@ -25,131 +24,116 @@
-  (model-ref [comm-atom]
-    "The COMM-ATOM's Jupyter Model reference identifier (a string).")
-  (origin-message [comm-atom]
-    "The message that caused the COMM-ATOM to be created.")
-  (state-set! [comm-atom comm-state]
-    "Sets the value `comm-atom` to be `comm-state` by updating the global state and sending COMM `update`
-  message.  `comm-state` must be a map serializable to JSON.  Returns `comm-atom`.")
-  (state-update! [comm-atom comm-state]
-    "Merges `comm-state` into current value of `comm-atom`.")
-  (watch [comm-atom key fn]
-    "Add watch to `comm-atom` which will subsequently be called when the value of `comm-atom` is
-    updated.  The observer `fn` fn must be a fn of 4 args: a key, the reference, its old-state, its
-    new-state, identical to those for `watch` functions (internally CommAtom uses an agent for
-    update notification), cf. `add-watch` for details.  Returns `comm-atom`.")
-  (unwatch [comm-atom key]
-    "Remove watch added with key `key` from `comm-atom`.  Returns `comm-atom`."))
-
-(declare agentfld comm-atom? send-comm-msg! send-comm-open! simple-fmt update-agent!)
+  (sync-state [comm-atom]
+    "A map of attributes to be sent to front-end")
+  (close! [comm-atom]
+    "Removes the comm-atom from the global state and sends COMM-CLOSE to the front end."))
+
+(declare comm-atom? send-comm-state! send-comm-open! simple-fmt jsonable?)
 
 (deftype CommAtom
-    [comm-state_ jup_ target-name_ reqmsg_ cid_ agent_]
+    [comm-id target sync-keys comm-state_]
 
   comm-atom-proto
-  (target [_]
-    target-name_)
-  (comm-id [_]
-    cid_)
-  (model-ref [_]
-    (str "IPY_MODEL_" cid_))
-  (origin-message [_]
-    reqmsg_)
-  (state-set! [comm-atom comm-state]
-    (assert (map? comm-state))
-    (reset! comm-state_ comm-state)
-    (update-agent! comm-atom comm-state)
-    (send-comm-msg! comm-atom comm-state)
-    comm-atom)
-  (state-update! [comm-atom comm-state]
-    (assert (map? comm-state))
-    (state-set! comm-atom (merge @comm-atom comm-state)))
-  (watch [comm-atom key f]
-    (assert (fn? f))
-    (add-watch (agentfld comm-atom) key f))
-  (unwatch [comm-atom key]
-    (remove-watch (agentfld comm-atom) key))
-
+  (sync-state [_]
+    (select-keys @comm-state_ sync-keys))
+  (close! [comm-atom]
+    (let [content (msgs/comm-close-content comm-id {})
+          jup (:jup @state/STATE)
+          {origin-message :req-message :or {origin-message {:header {}}}} (first (:cur-ctx @state/STATE))]
+      (jup/send!! jup :iopub_port origin-message msgs/COMM-CLOSE MESSAGE-METADATA content)
+      (swap! state/STATE update :comms dissoc comm-id)
+      nil))
+
+  java.io.Closeable
+  (close [comm-atom]
+    (msgs/leaf-paths comm-atom? #(.close %) (sync-state comm-atom))
+    (close! comm-atom))
 
   mc/PMimeConvertible
   (to-mime [_]
     (u/json-str {:text/plain
-                 ,, (str "[" cid_ "]=" @agent_)
+                 ,, (str "[" comm-id "]=" (:_model_name @comm-state_))
```

Comment:
Question why the `,,` prefix?

Reply:
No idea. I found it in the initial file and left it there since it has no tangible effect.